File: amino/backends/clickhouse.py
"""ClickHouse compile target. Emits server-side typed ``{name:Type}`` parameters."""

from typing import Any

from amino.errors import UnsupportedExpressionError

from .base import ParamSink, SQLBackend

_SCALAR_TYPES: dict[str, str] = {
    "Int": "Int64",
    "Float": "Float64",
    "Str": "String",
    "Bool": "Bool",
}


def _ch_type(base_type: str) -> str:
    if base_type.startswith("List[") and base_type.endswith("]"):
        return f"Array({_ch_type(base_type[5:-1])})"
    try:
        return _SCALAR_TYPES[base_type]
    except KeyError:
        raise UnsupportedExpressionError(f"No ClickHouse type for '{base_type}'") from None


class _NamedParams(ParamSink):
    def __init__(self) -> None:
        self._values: dict[str, Any] = {}

    def add(self, value: Any, base_type: str) -> str:
        name = f"p{len(self._values)}"
        self._values[name] = value
        return f"{{{name}:{_ch_type(base_type)}}}"

    def result(self) -> dict[str, Any]:
        return self._values


class ClickHouseBackend(SQLBackend):
    """Compile an expression to a ClickHouse ``WHERE`` predicate.

    ``Query.params`` is a dict for ``client.query(sql, parameters=params)``
    with clickhouse-connect. Placeholders carry the type inferred at parse
    time, so ClickHouse binds them server-side without string interpolation.
    """

    def quote_ident(self, name: str) -> str:
        return "`" + name.replace("\\", "\\\\").replace("`", "\\`") + "`"

    def new_params(self) -> ParamSink:
        return _NamedParams()

    def render_in(self, column: str, placeholder: str) -> str:
        return f"{column} IN {placeholder}"

    def render_not_in(self, column: str, placeholder: str) -> str:
        return f"{column} NOT IN {placeholder}"

    def render_contains(self, haystack: str, needle: str) -> str:
        return f"position({haystack}, {needle}) > 0"

File: go/backend/clickhouse.go
package backend

import (
	"strings"

	amino "github.com/raiderrobert/amino/go"
)

// ClickHouse compiles to a predicate with server-side typed {p0:Int64}
// placeholders. Params is a map[string]any for the clickhouse-go client's
// named parameters. The type comes from the schema, so nothing is guessed
// from the value at bind time.
type ClickHouse struct {
	Columns map[string]string
}

// Compile renders expr for ClickHouse.
func (c ClickHouse) Compile(expr *amino.Expression) (Query, error) {
	return Compile(chDialect{}, expr, c.Columns)
}

var chScalar = map[string]string{"Int": "Int64", "Float": "Float64", "Str": "String", "Bool": "Bool"}

func chType(base string) string {
	if strings.HasPrefix(base, "List[") && strings.HasSuffix(base, "]") {
		return "Array(" + chType(base[5:len(base)-1]) + ")"
	}
	if t, ok := chScalar[base]; ok {
		return t
	}
	return "String"
}

type chDialect struct{}

type chParams struct{ values map[string]any }

func (s *chParams) Add(value any, base string) string {
	if s.values == nil {
		s.values = map[string]any{}
	}
	name := "p" + itoa(len(s.values))
	s.values[name] = value
	return "{" + name + ":" + chType(base) + "}"
}

func (s *chParams) Result() any {
	if s.values == nil {
		return map[string]any{}
	}
	return s.values
}

func (chDialect) Name() string { return "ClickHouse" }
func (chDialect) QuoteIdent(name string) string {
	return "`" + strings.ReplaceAll(strings.ReplaceAll(name, `\`, `\\`), "`", "\\`") + "`"
}
func (chDialect) NewParams() ParamSink              { return &chParams{} }
func (chDialect) RenderIn(col, ph string) string    { return col + " IN " + ph }
func (chDialect) RenderNotIn(col, ph string) string { return col + " NOT IN " + ph }
func (chDialect) RenderContains(hay, needle string) string {
	return "position(" + hay + ", " + needle + ") > 0"
}
